src/lib.rs

```diff
@@ -1,32 +1,305 @@
+// SYSTEM CONSTS
+//--------------------------------------------------------------------------------------------------
+
+// HC.Version
+const VERSION : u16 = 1;
+const VERSION_STR : &'static str = "1";
+
+
+// HC.HashNotFound
+// FIXME keep in sync with HcApiReturnCode?
+pub enum ErrorCode {
+  FunctionNotImplemented,
+  HashNotFound,
+}
+
 
-type Hash = String;
+// HC.Status
+// WARNING keep in sync with CRUDStatus
+bitflags! {
+  pub struct EntryStatus: u8 {
+    const LIVE     = 1 << 0;
+    const REJECTED = 1 << 1;
+    const DELETED  = 1 << 2;
+    const MODIFIED = 1 << 3;
+  }
+}
 
-pub mod anchors;
 
-pub fn commit<S: Into<String>>(entry_type: S, entry: serde_json::Value) -> String {
-    "".to_string()
+// HC.GetMask
+bitflags! {
+  pub struct GetMask: u8 {
+    const ENTRY      = 1 << 0;
+    const ENTRY_TYPE = 1 << 1;
+    const SOURCES    = 1 << 2;
+  }
+}
+// explicit `Default` implementation
+impl Default for GetMask {
+    fn default() -> GetMask {
+        GetMask::ENTRY
+    }
 }
 
-pub fn link<S: Into<String>>(base: Hash, tag: S, target: Hash) {
 
+// HC.LinkAction
+pub enum LinkAction {
+    ADD,
+    DELETE,
 }
 
-pub fn get(entry_hash: Hash) -> Option<String> {
-    None
+
+// HC.PkgReq
+pub enum PkgRequest {
+    CHAIN,
+    CHAIN_OPTION,
+    ENTRY_TYPES,
 }
 
-pub fn get_links<S: Into<String>>(bash: Hash, tag: S) -> Vec<Hash> {
-    vec![]
+
+// HC.PkgReq.ChainOpt
+pub enum ChainOption {
+    NONE,
+    HEADERS,
+    ENTRIES,
+    FULL,
+}
+
+
+// HC.Bridge
+pub enum BridgeSide {
+    FROM,
+    TO,
+}
+
+
+// HC.SysEntryType
+enum SystemEntryType {
+    DNA,
+    AGENT,
+    KEY,
+    HEADERS,
+    DELETION,
 }
 
 
+mod bundle_cancel {
+    // HC.BundleCancel.Reason
+    pub enum Reason {
+        USER_CANCEL,
+        TIMEOUT,
+    }
+    // HC.BundleCancel.Response
+    pub enum Response {
+        OK,
+        COMMIT,
+    }
+}
+
+
+/// Allowed input for close_bundle()
+pub enum BundleOnClose {
+    Commit,
+    Discard,
+}
+
+
+//--------------------------------------------------------------------------------------------------
+// API FUNCTIONS
+//--------------------------------------------------------------------------------------------------
+
+/// FIXME DOC
+/// Returns an application property, which are defined by the app developer.
+/// It returns values from the DNA file that you set as properties of your application
+/// (e.g. Name, Language, Description, Author, etc.).
+pub fn property<S: Into<String>>(_name: S) -> Result<String, ErrorCode> {
+    // FIXME
+    Err(ErrorCode::FunctionNotImplemented)
+}
+
+/// FIXME DOC
+pub fn make_hash<S: Into<String>>(_entry_type : S, _entry_data: serde_json::Value) -> Result<HashString, ErrorCode> {
+    // FIXME
+    Err(ErrorCode::FunctionNotImplemented)
+}
 
 
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+/// FIXME DOC
+pub fn debug(msg: &str) {
+    unsafe {
```

Comment:
This shouldn't span over too much of the code. Could move this down, so that only the unsafe function call is contained?

Reply:
Fixed but `g_mem_stack` is a mutable static, which is also unsafe to use.